From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Sun, 19 Sep 2021 13:11:13 +0200
Subject: [PATCH] Add MongoDB


diff --git a/build.gradle.kts b/build.gradle.kts
index 241ee5c98275a4517e040c86355ae6702f8efca1..dd91756ce7501c68c55afcf7c948431450bb5506 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -23,6 +23,7 @@ configurations.api {
 
 dependencies {
     // api dependencies are listed transitively to API consumers
+    api("org.mongodb:mongo-java-driver:3.12.10") // Yokura
     api("commons-lang:commons-lang:2.6")
     api("com.google.guava:guava:31.0.1-jre")
     api("com.google.code.gson:gson:2.8.8")
diff --git a/src/main/java/fr/hygon/yokura/IYokuraAPI.java b/src/main/java/fr/hygon/yokura/IYokuraAPI.java
new file mode 100644
index 0000000000000000000000000000000000000000..ff3658c487a52502dc9a529b3d92a67eb8cc7513
--- /dev/null
+++ b/src/main/java/fr/hygon/yokura/IYokuraAPI.java
@@ -0,0 +1,16 @@
+package fr.hygon.yokura;
+
+import com.mongodb.MongoClient;
+import com.mongodb.client.MongoDatabase;
+
+public interface IYokuraAPI {
+    /**
+     * @return the mongo client instance
+     */
+    MongoClient getMongoClient();
+
+    /**
+     * @return the default database used by the server
+     */
+    MongoDatabase getMongoDatabase();
+}
\ No newline at end of file
diff --git a/src/main/java/fr/hygon/yokura/YokuraAPI.java b/src/main/java/fr/hygon/yokura/YokuraAPI.java
new file mode 100644
index 0000000000000000000000000000000000000000..ea3204863ee6681c70e7735dc4aa45f8903ea5b8
--- /dev/null
+++ b/src/main/java/fr/hygon/yokura/YokuraAPI.java
@@ -0,0 +1,44 @@
+package fr.hygon.yokura;
+
+import com.mongodb.MongoClient;
+import com.mongodb.client.MongoDatabase;
+
+public class YokuraAPI {
+    private static IYokuraAPI instance;
+
+    /**
+     * @return the instance of the API
+     */
+    public static IYokuraAPI getInstance() {
+        return instance;
+    }
+
+    /**
+     * Attempts to set the {@link IYokuraAPI} singleton.
+     * <p>
+     * Cannot be done if the instance is already set.
+     *
+     * @param instance IYokuraAPI instance
+     */
+    public static void setInstance(IYokuraAPI instance) {
+        if(YokuraAPI.instance != null) {
+            throw new UnsupportedOperationException("Cannot redefine singleton YokuraAPI");
+        }
+
+        YokuraAPI.instance = instance;
+    }
+
+    /**
+     * @return the mongo client instance
+     */
+    public static MongoClient getMongoClient() {
+        return instance.getMongoClient();
+    }
+
+    /**
+     * @return the default database used by the server
+     */
+    public static MongoDatabase getMongoDatabase() {
+        return instance.getMongoDatabase();
+    }
+}
