From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Mon, 20 Sep 2021 19:01:44 +0200
Subject: [PATCH] Add broker


diff --git a/build.gradle.kts b/build.gradle.kts
index dd91756ce7501c68c55afcf7c948431450bb5506..6a2d184cbf902e8d7dc0de5a2b09df90892620de 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -21,9 +21,17 @@ configurations.api {
     extendsFrom(apiAndDocs)
 }
 
+
+repositories {
+    maven("http://37.59.90.35/maven/") {
+        isAllowInsecureProtocol = true
+    }
+}
+
 dependencies {
     // api dependencies are listed transitively to API consumers
     api("org.mongodb:mongo-java-driver:3.12.10") // Yokura
+    api("fr.hygon:brokerapi:1.0") // Yokura
     api("commons-lang:commons-lang:2.6")
     api("com.google.guava:guava:31.0.1-jre")
     api("com.google.code.gson:gson:2.8.8")
diff --git a/src/main/java/fr/hygon/yokura/IYokuraAPI.java b/src/main/java/fr/hygon/yokura/IYokuraAPI.java
index ff3658c487a52502dc9a529b3d92a67eb8cc7513..d8592af9cc94fc4337e471687f1bf64505f8783e 100644
--- a/src/main/java/fr/hygon/yokura/IYokuraAPI.java
+++ b/src/main/java/fr/hygon/yokura/IYokuraAPI.java
@@ -2,6 +2,7 @@ package fr.hygon.yokura;
 
 import com.mongodb.MongoClient;
 import com.mongodb.client.MongoDatabase;
+import fr.hygon.brokerapi.packets.Packet;
 
 public interface IYokuraAPI {
     /**
@@ -13,4 +14,10 @@ public interface IYokuraAPI {
      * @return the default database used by the server
      */
     MongoDatabase getMongoDatabase();
+
+    /**
+     * Sends a packet to the broker that will be re-sent to <b>all</b> the connected clients.
+     * @param packet the packet that will be sent.
+     */
+    void sendPacket(Packet packet);
 }
\ No newline at end of file
diff --git a/src/main/java/fr/hygon/yokura/YokuraAPI.java b/src/main/java/fr/hygon/yokura/YokuraAPI.java
index ea3204863ee6681c70e7735dc4aa45f8903ea5b8..9d9330bdd70eed77dae5e72b829fd32164b4a4db 100644
--- a/src/main/java/fr/hygon/yokura/YokuraAPI.java
+++ b/src/main/java/fr/hygon/yokura/YokuraAPI.java
@@ -2,6 +2,7 @@ package fr.hygon.yokura;
 
 import com.mongodb.MongoClient;
 import com.mongodb.client.MongoDatabase;
+import fr.hygon.brokerapi.packets.Packet;
 
 public class YokuraAPI {
     private static IYokuraAPI instance;
@@ -41,4 +42,12 @@ public class YokuraAPI {
     public static MongoDatabase getMongoDatabase() {
         return instance.getMongoDatabase();
     }
+
+    /**
+     * Sends a packet to the broker that will be re-sent to <b>all</b> the connected clients.
+     * @param packet the packet that will be sent.
+     */
+    public void sendPacket(Packet packet) {
+        instance.sendPacket(packet);
+    }
 }
