From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Sun, 10 Oct 2021 19:10:56 +0200
Subject: [PATCH] Add the server system


diff --git a/src/main/java/fr/hygon/yokura/IYokuraAPI.java b/src/main/java/fr/hygon/yokura/IYokuraAPI.java
index d8592af9cc94fc4337e471687f1bf64505f8783e..2f0c5764eac90c879cdf41467c987bddb429463a 100644
--- a/src/main/java/fr/hygon/yokura/IYokuraAPI.java
+++ b/src/main/java/fr/hygon/yokura/IYokuraAPI.java
@@ -3,6 +3,7 @@ package fr.hygon.yokura;
 import com.mongodb.MongoClient;
 import com.mongodb.client.MongoDatabase;
 import fr.hygon.brokerapi.packets.Packet;
+import fr.hygon.brokerapi.packets.ServerStatusPacket;
 
 public interface IYokuraAPI {
     /**
@@ -20,4 +21,16 @@ public interface IYokuraAPI {
      * @param packet the packet that will be sent.
      */
     void sendPacket(Packet packet);
+
+    /**
+     * @return the server id. Returns null if the current server is a static server.
+     */
+    String getServerID();
+
+    /**
+     * Updates the status of the server
+     * @param status the nex status
+     * @see ServerStatusPacket.Status for the status list
+     */
+    void setStatus(ServerStatusPacket.Status status);
 }
\ No newline at end of file
diff --git a/src/main/java/fr/hygon/yokura/YokuraAPI.java b/src/main/java/fr/hygon/yokura/YokuraAPI.java
index 9d9330bdd70eed77dae5e72b829fd32164b4a4db..8009f188281e0ce24c0982ef2e205978d26d8567 100644
--- a/src/main/java/fr/hygon/yokura/YokuraAPI.java
+++ b/src/main/java/fr/hygon/yokura/YokuraAPI.java
@@ -3,6 +3,7 @@ package fr.hygon.yokura;
 import com.mongodb.MongoClient;
 import com.mongodb.client.MongoDatabase;
 import fr.hygon.brokerapi.packets.Packet;
+import fr.hygon.brokerapi.packets.ServerStatusPacket;
 
 public class YokuraAPI {
     private static IYokuraAPI instance;
@@ -50,4 +51,21 @@ public class YokuraAPI {
     public void sendPacket(Packet packet) {
         instance.sendPacket(packet);
     }
+
+
+    /**
+     * @return the server id. Returns null if the current server is a static server.
+     */
+    public static String getServerID() {
+        return instance.getServerID();
+    }
+
+    /**
+     * Updates the status of the server
+     * @param status the nex status
+     * @see ServerStatusPacket.Status for the status list
+     */
+    public static void setStatus(ServerStatusPacket.Status status) {
+        instance.setStatus(status);
+    }
 }
diff --git a/src/main/java/fr/hygon/yokura/servers/Status.java b/src/main/java/fr/hygon/yokura/servers/Status.java
new file mode 100644
index 0000000000000000000000000000000000000000..edd1fd651b866d02b0d299be255ae31790edd523
--- /dev/null
+++ b/src/main/java/fr/hygon/yokura/servers/Status.java
@@ -0,0 +1,8 @@
+package fr.hygon.yokura.servers;
+
+public enum Status {
+    STARTING(),
+    WAITING_PLAYERS(),
+    IN_GAME(),
+    STOPPING();
+}
