From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Sun, 19 Sep 2021 17:12:25 +0200
Subject: [PATCH] Add network levels and xp


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index c699bce3f87295a369af98b0b3e39576956deed4..1fec543f820130b25d63314b1b9de75f19340bed 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -267,6 +267,9 @@ public class ServerPlayer extends Player {
     // Yokura start
     public double coins;
     public double tokens;
+
+    public int networkXP;
+    public int networkLevel;
     // Yokura end
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile, ServerLoginPacketListenerImpl loginPacketListener) { // Yokura - add loginPacketListener
@@ -343,6 +346,9 @@ public class ServerPlayer extends Player {
         // Yokura start
         coins = loginPacketListener.coins;
         tokens = loginPacketListener.tokens;
+
+        networkXP = loginPacketListener.networkXP;
+        networkLevel = loginPacketListener.networkLevel;
         // Yokura end
     }
     // Paper start - Chunk priority
diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 076f1e16d1484a299205796fbaa8ae3790757b4d..f5bdff50d25b4f08c7a20b708e6ef16d2c01728a 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -73,6 +73,9 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener
     // Yokura start
     public double coins;
     public double tokens;
+
+    public int networkXP;
+    public int networkLevel;
     // Yokura end
 
     public ServerLoginPacketListenerImpl(MinecraftServer server, Connection connection) {
@@ -469,17 +472,22 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener
 
             playerData = new Document("_id", playerUUID.toString())
                     .append("coins", YokuraGlobalConfig.defaultCoins)
-                    .append("tokens", YokuraGlobalConfig.defaultTokens);
+                    .append("tokens", YokuraGlobalConfig.defaultTokens)
+                    .append("xp", 0)
+                    .append("level", 0);
 
             coins = YokuraGlobalConfig.defaultCoins;
             tokens = YokuraGlobalConfig.defaultTokens;
+            networkXP = 0;
+            networkLevel = 0;
 
             YokuraAPI.getMongoDatabase().getCollection("players").insertOne(playerData);
-
             ServerLoginPacketListenerImpl.LOGGER.info("Profile for the player " + playerName + " created! (Took " + (System.currentTimeMillis() - startTime) + "ms).");
         } else {
             coins = playerData.getDouble("coins");
             tokens = playerData.getDouble("tokens");
+            networkXP = playerData.getInteger("xp");
+            networkLevel = playerData.getInteger("level");
         }
     }
     // Yokura end
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index c29d70aa855e3580733f2b1ab57827fd95021cd3..83ab92461fba466a512428e077532896a9a2cf46 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -88,6 +88,38 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     public boolean hasEnoughTokens(double tokens) {
         return getTokens() >= tokens;
     }
+
+    @Override
+    public int getNetworkXp() {
+        return MongoUtils.getInt("players", getUniqueId().toString(), "xp");
+    }
+
+    @Override
+    public int getNetworkLevel() {
+        return MongoUtils.getInt("players", getUniqueId().toString(), "level");
+    }
+
+    @Override
+    public void addNetworkXp(int xp) {
+        if(xp <= 0) return;
+        java.util.concurrent.CompletableFuture.runAsync(() -> {
+           int networkXp = getNetworkXp();
+           int networkLevel = getNetworkLevel();
+
+           int newXp = networkXp + xp;
+           if(newXp >= 250) {
+               while (newXp >= 250) {
+                   newXp = newXp - 250;
+                   networkLevel++;
+               }
+
+               new fr.hygon.yokura.event.player.NetworkLevelChangeEvent(this, networkLevel, newXp).callEvent();
+               MongoUtils.updateValue("players", getUniqueId().toString(), "level", networkLevel);
+           }
+
+           MongoUtils.updateValue("players", getUniqueId().toString(), "xp", newXp);
+        });
+    }
     // Yokura end
 
     public GameProfile getProfile() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 2db6c2542181694f6c70ac6dce67ab79e5ec9337..d50e41ca8cf6e15fb5bf0234569a7f735ce7d333 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -221,6 +221,35 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public boolean hasEnoughTokens(double tokens) {
         return getTokens() >= tokens;
     }
+
+    @Override
+    public int getNetworkXp() {
+        return getHandle().networkXP;
+    }
+
+    @Override
+    public int getNetworkLevel() {
+        return getHandle().networkLevel;
+    }
+
+    @Override
+    public void addNetworkXp(int xp) {
+        if(xp <= 0) return;
+        int networkLevel = getNetworkLevel();
+
+        int newXp = getNetworkXp() + xp;
+        if(newXp >= 250) {
+            while (newXp >= 250) {
+                newXp = newXp - 250;
+                networkLevel++;
+            }
+
+            new fr.hygon.yokura.event.player.NetworkLevelChangeEvent(this, networkLevel, newXp).callEvent();
+            MongoUtils.updateValueAsync("players", getUniqueId().toString(), "level", networkLevel);
+        }
+
+        MongoUtils.updateValueAsync("players", getUniqueId().toString(), "xp", newXp);
+    }
     // Yokura end
 
     public GameProfile getProfile() {
