From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Mon, 20 Sep 2021 19:01:36 +0200
Subject: [PATCH] Add broker


diff --git a/build.gradle.kts b/build.gradle.kts
index 3b8fafbb37811d248e2544012684063c5ae811e2..c80fceaa940236c6b03861154faf3f57182f3981 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -15,6 +15,10 @@ repositories {
         mavenContent { includeModule("net.fabricmc", "mapping-io") }
     }
     // Paper end
+
+    maven("http://37.59.90.35/maven/") {
+        isAllowInsecureProtocol = true
+    }
 }
 
 dependencies {
diff --git a/src/main/java/fr/hygon/yokura/YokuraAPIServer.java b/src/main/java/fr/hygon/yokura/YokuraAPIServer.java
index 6e403a6d5f7990478c5bc372b75a68398bf8eee3..a4fcd77007532725ea1e60359d4830dc932d635c 100644
--- a/src/main/java/fr/hygon/yokura/YokuraAPIServer.java
+++ b/src/main/java/fr/hygon/yokura/YokuraAPIServer.java
@@ -2,16 +2,10 @@ package fr.hygon.yokura;
 
 import com.mongodb.MongoClient;
 import com.mongodb.client.MongoDatabase;
+import fr.hygon.brokerapi.Broker;
+import fr.hygon.brokerapi.packets.Packet;
 
-public class YokuraAPIServer implements IYokuraAPI {
-    private final MongoClient mongoClient;
-    private final MongoDatabase mongoDatabase;
-
-    public YokuraAPIServer(MongoClient mongoClient, MongoDatabase mongoDatabase) {
-        this.mongoClient = mongoClient;
-        this.mongoDatabase = mongoDatabase;
-    }
-
+public record YokuraAPIServer(MongoClient mongoClient, MongoDatabase mongoDatabase, Broker broker) implements IYokuraAPI {
     @Override
     public MongoClient getMongoClient() {
         return mongoClient;
@@ -21,4 +15,9 @@ public class YokuraAPIServer implements IYokuraAPI {
     public MongoDatabase getMongoDatabase() {
         return mongoDatabase;
     }
+
+    @Override
+    public void sendPacket(Packet packet) {
+        broker.sendPacket(packet);
+    }
 }
diff --git a/src/main/java/fr/hygon/yokura/broker/PacketListenerImpl.java b/src/main/java/fr/hygon/yokura/broker/PacketListenerImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..70bd0a40c3d784f685cd5f32df3e124b05f75cea
--- /dev/null
+++ b/src/main/java/fr/hygon/yokura/broker/PacketListenerImpl.java
@@ -0,0 +1,17 @@
+package fr.hygon.yokura.broker;
+
+import fr.hygon.brokerapi.packets.PacketListener;
+import fr.hygon.brokerapi.packets.ServerStatusPacket;
+import fr.hygon.brokerapi.packets.VelocityKeyUpdatePacket;
+
+public class PacketListenerImpl implements PacketListener {
+    @Override
+    public void handleVelocityKeyUpdate(VelocityKeyUpdatePacket velocityKeyUpdatePacket) {
+
+    }
+
+    @Override
+    public void handleServerStatusUpdate(ServerStatusPacket serverStatusPacket) {
+
+    }
+}
diff --git a/src/main/java/fr/hygon/yokura/config/YokuraGlobalConfig.java b/src/main/java/fr/hygon/yokura/config/YokuraGlobalConfig.java
index b92672eebf9b815c51552651cc2dee7cfbc3abfb..7547132521b586f329abe0d51c89b4245d12aa1f 100644
--- a/src/main/java/fr/hygon/yokura/config/YokuraGlobalConfig.java
+++ b/src/main/java/fr/hygon/yokura/config/YokuraGlobalConfig.java
@@ -113,6 +113,16 @@ public class YokuraGlobalConfig {
         mongoPassword = getString("credentials.mongodb.password", "").toCharArray();
     }
 
+    public static String brokerHost = "localhost";
+    private static void getBrokerHost() {
+        brokerHost = getString("credentials.broker.host", brokerHost);
+    }
+
+    public static int brokerPort = 9800;
+    private static void getBrokerPort() {
+        brokerPort = getInt("credentials.broker.port", brokerPort);
+    }
+
     public static double defaultCoins = 200.0;
     private static void getDefaultCoins() {
         defaultCoins = getDouble("players.default-coins", defaultCoins);
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 056972869403c41b80a56f57fbc802a1f78853b6..04ab2817f2ff1a8dd406d4008402041766fd272a 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -21,6 +21,7 @@ import java.util.Locale;
 import java.util.function.BooleanSupplier;
 import java.util.regex.Pattern;
 import javax.annotation.Nullable;
+import fr.hygon.brokerapi.Broker; // Yokura
 import fr.hygon.yokura.config.YokuraGlobalConfig; // Yokura
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
@@ -254,7 +255,16 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             return false;
         }
 
-        fr.hygon.yokura.YokuraAPI.setInstance(new fr.hygon.yokura.YokuraAPIServer(mongoClient, mongoDatabase));
+        Broker broker = new Broker(YokuraGlobalConfig.brokerHost, YokuraGlobalConfig.brokerPort, new fr.hygon.yokura.broker.PacketListenerImpl());
+
+        try {
+            broker.connect();
+            DedicatedServer.LOGGER.info("Connected to the broker.");
+        } catch (InterruptedException exception) {
+            DedicatedServer.LOGGER.error("Couldn't connect to the broker.", exception);
+            return false;
+        }
+        fr.hygon.yokura.YokuraAPI.setInstance(new fr.hygon.yokura.YokuraAPIServer(mongoClient, mongoDatabase, broker));
         // Yokura end
 
         this.setPvpAllowed(dedicatedserverproperties.pvp);
