From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Wed, 20 Oct 2021 13:38:57 +0200
Subject: [PATCH] Add QuestionPrompt utility


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index b11066b9b75eb89d7eb35ea308b671e0311b5186..d9b8582b50ed90dbf243e603f8eab62a4efe119f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -273,6 +273,8 @@ public class ServerPlayer extends Player {
 
     public ParticleList particle;
     public ArrayList<ParticleList> boughtParticles;
+
+    @Nullable public fr.hygon.yokura.entity.player.QuestionPrompt questionPrompt = null;
     // Yokura end
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile, ServerLoginPacketListenerImpl loginPacketListener) { // Yokura - add loginPacketListener
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 057fcbc389e54e0c9f7a90a3e8b965cd46db9d58..1e907db538f3379d3e69689b050183edfa3c8e7b 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1961,6 +1961,12 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
             this.processedDisconnect = true;
         }
         // CraftBukkit end
+        // Yokura start
+        if(this.player.questionPrompt != null) {
+            this.player.questionPrompt.onQuit();
+            this.player.questionPrompt = null;
+        }
+        // Yokura end
         ServerGamePacketListenerImpl.LOGGER.info("{} lost connection: {}", this.player.getName().getString(), reason.getString());
         // CraftBukkit start - Replace vanilla quit message handling with our own.
         /*
@@ -2585,7 +2591,14 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         PacketUtils.ensureRunningOnSameThread(packetplayinclosewindow, this, this.player.getLevel());
 
         if (this.player.isImmobile()) return; // CraftBukkit
-        CraftEventFactory.handleInventoryCloseEvent(this.player, reason); // CraftBukkit // Paper
+        // Yokura start
+        if(this.player.questionPrompt != null) {
+            this.player.questionPrompt.onQuit();
+            this.player.questionPrompt = null;
+        } else {
+            CraftEventFactory.handleInventoryCloseEvent(this.player, reason); // CraftBukkit // Paper
+        }
+        // Yokura end
 
         this.player.doCloseContainer();
     }
@@ -2786,6 +2799,26 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
                     if (click == ClickType.NUMBER_KEY) {
                         event = new InventoryClickEvent(inventory, type, packet.getSlotNum(), click, action, packet.getButtonNum());
                     } else {
+                        // Yokura start
+                        if(this.player.questionPrompt != null) {
+                            org.bukkit.inventory.ItemStack clickedItem = inventory.getItem(packet.getSlotNum());
+                            if(clickedItem != null) {
+                                if(clickedItem.getType() == org.bukkit.Material.GREEN_STAINED_GLASS_PANE) {
+                                    this.player.questionPrompt.onAccept();
+                                    this.player.questionPrompt = null;
+                                    this.player.closeContainer();
+                                } else if(clickedItem.getType() == org.bukkit.Material.RED_STAINED_GLASS_PANE) {
+                                    this.player.questionPrompt.onRefuse();
+                                    this.player.questionPrompt = null;
+                                    this.player.closeContainer();
+                                } else if(clickedItem.getType() == org.bukkit.Material.BARRIER) {
+                                    this.player.questionPrompt.onQuit();
+                                    this.player.questionPrompt = null;
+                                    this.player.closeContainer();
+                                }
+                            }
+                        }
+                        // Yokura end
                         event = new InventoryClickEvent(inventory, type, packet.getSlotNum(), click, action);
                     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 87dcbb4fce2fdcd648da96499b5cf55e7ebe8607..407f6db6bdb6fc9cd60dc293452e93d084f5ce9b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -313,6 +313,42 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public @NotNull List<ParticleList> getBoughtParticles() {
         return getHandle().boughtParticles;
     }
+
+    @Override
+    public void askQuestion(net.kyori.adventure.text.Component title, fr.hygon.yokura.entity.player.QuestionPrompt questionPrompt) {
+        org.bukkit.inventory.Inventory GUI = Bukkit.createInventory(null, 27, title);
+
+        ItemStack accept = new ItemStack(Material.GREEN_STAINED_GLASS_PANE, 1);
+        org.bukkit.inventory.meta.ItemMeta acceptMeta = accept.getItemMeta();
+
+        acceptMeta.displayName(net.kyori.adventure.text.Component.text("Accepter").color(net.kyori.adventure.text.format.TextColor.color(0, 140, 0))
+                .decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false));
+        accept.setItemMeta(acceptMeta);
+
+        ItemStack refuse = new ItemStack(Material.RED_STAINED_GLASS_PANE, 1);
+        org.bukkit.inventory.meta.ItemMeta refuseMeta = accept.getItemMeta();
+
+        refuseMeta.displayName(net.kyori.adventure.text.Component.text("Refuser").color(net.kyori.adventure.text.format.TextColor.color(225, 45, 40))
+                .decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false));
+        refuse.setItemMeta(refuseMeta);
+
+        ItemStack cancel = new ItemStack(Material.BARRIER, 1);
+        org.bukkit.inventory.meta.ItemMeta cancelMeta = accept.getItemMeta();
+
+        cancelMeta.displayName(net.kyori.adventure.text.Component.text("Annuler").color(net.kyori.adventure.text.format.TextColor.color(180, 0, 0))
+                .decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false));
+        cancel.setItemMeta(cancelMeta);
+
+        GUI.setItem(11, accept);
+        GUI.setItem(15, refuse);
+        GUI.setItem(22, cancel);
+
+        questionPrompt.onAsk();
+        closeInventory();
+        openInventory(GUI);
+        getHandle().questionPrompt = questionPrompt; // We must set it AFTER we opened the inventory,
+        // else the quit method will be called if the player had a GUI opened before because it'll close the old GUI and open this one
+    }
     // Yokura end
 
     public GameProfile getProfile() {
