From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Tue, 26 Oct 2021 21:16:56 +0200
Subject: [PATCH] Add an API for the server system


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 407f6db6bdb6fc9cd60dc293452e93d084f5ce9b..9915efa054bad7073afc67745c5a9cef948b6f5b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -11,6 +11,7 @@ import fr.hygon.yokura.entity.player.ParticleList; // Yokura
 import fr.hygon.yokura.entity.player.Ranks; // Yokura
 import io.netty.buffer.Unpooled;
 import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream; // Yokura
 import java.io.IOException;
 import java.lang.ref.WeakReference;
 import java.net.InetSocketAddress;
@@ -349,6 +350,22 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().questionPrompt = questionPrompt; // We must set it AFTER we opened the inventory,
         // else the quit method will be called if the player had a GUI opened before because it'll close the old GUI and open this one
     }
+
+    @Override
+    public void sendToServer(@NotNull String server) {
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
+        try {
+            dataOutputStream.writeUTF("Connect");
+            dataOutputStream.writeUTF(server);
+        } catch (IOException exception) {
+            exception.printStackTrace();
+        }
+
+        ClientboundCustomPayloadPacket packet = new ClientboundCustomPayloadPacket(new ResourceLocation("bungeecord:main"),
+            new FriendlyByteBuf(Unpooled.wrappedBuffer(byteArrayOutputStream.toByteArray())));
+        getHandle().connection.send(packet);
+    }
     // Yokura end
 
     public GameProfile getProfile() {
